### PR TITLE
Feature/tp2squadmate

### PR DESCRIPTION
**Pull request by Crowdedlight**
Pull request adds the Teleport To Squadmate feature together with some helpers used in this feature:

Teleport to squadmate allow GM to right-click a player and select what squadmate to teleport the player to. Makes it easy to move a lost or respawned player back to his squad.

1. Adds ListBox dialog GUI. Can be re-used if a single-select listbox is needed elsewhere.
2. Adds loghelper, to try and standardize format for error prints. (Only this PR is using is. If you agree it makes sense I can do another PR after this, that changes existing errorprints to use it). 
3. Adds GroupHelper to get group from player entity. (Thanks Kex! Using your RSA helper)

This feature does not work on AI or AI groups as it would require specific handling and at this point, AI can't be in groups with players anyway. 

billede
billede

### CODE REVIEW

**Review by Kexanone — CHANGES_REQUESTED** [review]
Trying to use this action while a team mate is in the respawn menu, produces a script error.

**Comment by Crowdedlight** [comment]
> Trying to use this action while a team mate is in the respawn menu, produces a script error.

I had forgotten a null check for dead players. Added now. Although if all squadmembers are dead now, so the list is empty, it just shows the empty list. Maybe would be better if empty list, to close the list and show hint to GM that there is no valid squadmembers to TP to?

**Comment by Kexanone** [comment]
> > Trying to use this action while a team mate is in the respawn menu, produces a script error.
> 
> I had forgotten a null check for dead players. Added now. Although if all squadmembers are dead now, so the list is empty, it just shows the empty list. Maybe would be better if empty list, to close the list and show hint to GM that there is no valid squadmembers to TP to?

I would say either way is fine.

**Comment by Crowdedlight** [comment]
> > > Trying to use this action while a team mate is in the respawn menu, produces a script error.
> > 
> > 
> > I had forgotten a null check for dead players. Added now. Although if all squadmembers are dead now, so the list is empty, it just shows the empty list. Maybe would be better if empty list, to close the list and show hint to GM that there is no valid squadmembers to TP to?
> 
> I would say either way is fine.

I think I am just going to leave it be an empty menu for now. Only in rare cases it would happen. The other approach would require making some hint helper, to send localized hints to specific players over Rpl. As the context-action is run with ``IsServer`` so would need to somehow get dynamic playerID of the current zeus, and then rpc the show-hint function to that player_id. And not quite sure how to actually get current zeus, from a function run on the server.

File: Scripts/Game/ODIN/Editor/Containers/Actions/ContextActions/ODIN_TeleportToSquadmateContextAction.c [hunk 1]
```diff
@@ -0,0 +1,134 @@
+//------------------------------------------------------------------------------------------------
+[BaseContainerProps(), SCR_BaseContainerCustomTitleUIInfo("m_Info")]
+class ODIN_TeleportToSquadmateContextAction: SCR_BaseContextAction
+{	
+	override bool CanBeShown(SCR_EditableEntityComponent hoveredEntity, notnull set<SCR_EditableEntityComponent> selectedEntities, vector cursorWorldPosition, int flags)
+	{
+		if (!hoveredEntity)
+			return false;
+		
+		GenericEntity owner = hoveredEntity.GetOwner();
```
Comment (Kexanone) [inline Scripts/Game/ODIN/Editor/Containers/Actions/ContextActions/ODIN_TeleportToSquadmateContextAction.c:10]: You could merge this with the cast to `SCR_ChimeraCharacter` check below

File: Scripts/Game/ODIN/Editor/Containers/Actions/ContextActions/ODIN_TeleportToSquadmateContextAction.c [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+			// TODO give error message back to zeus that unit is not in a group 
+			ODIN_LogHelper.Log("null group or <= 0 in group", "TeleportToSquadmateContextAction", LogLevel.WARNING);
+			return;
+		}
+		
+		// can't get character entities directly from group, only for AI. So for players we get their IDs and gotta find and cast them with that
+		array<int> members = group.GetPlayerIDs();
+		foreach (int playerID : members)
+		{
+			PlayerManager playerManager = GetGame().GetPlayerManager();
```
Comment (Kexanone) [inline Scripts/Game/ODIN/Editor/Containers/Actions/ContextActions/ODIN_TeleportToSquadmateContextAction.c:65]: Get `PlayerManager` outside of the loop

File: Scripts/Game/ODIN/Editor/Containers/Actions/ContextActions/ODIN_TeleportToSquadmateContextAction.c [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+			string name = playerManager.GetPlayerName(playerID);
+			ChimeraCharacter player = ChimeraCharacter.Cast(playerManager.GetPlayerControlledEntity(playerID));
+			
+			// if player is dead and the object is null, we skip
+			if (!player)
+				continue;
+			
+			// don't add the unit we want to teleport
+			if (player.GetID() == owner.GetID())
```
Comment (Kexanone) [inline Scripts/Game/ODIN/Editor/Containers/Actions/ContextActions/ODIN_TeleportToSquadmateContextAction.c:74]: I think comparing `player` with `owner` directly would also work

File: Scripts/Game/ODIN/Editor/Containers/Actions/ContextActions/ODIN_TeleportToSquadmateContextAction.c [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+		ChimeraCharacter selectedPlayer = ChimeraCharacter.Cast(listbox.GetItemData(selectedItem));
+		if (!selectedPlayer)
+		{
+			ODIN_LogHelper.Log("ChimeraCharacter in userdata for selected listbox item is null", "TeleportToSquadmateContextAction", LogLevel.WARNING);
+			return;
+		}
+
+		//if (selectedPlayer.IsInVehicle())
+		//{
+			// try to put hoveredentity inside vehicle? otherwise keep going and put it next to vic
```
Comment (Kexanone) [inline Scripts/Game/ODIN/Editor/Containers/Actions/ContextActions/ODIN_TeleportToSquadmateContextAction.c:112]: Check out `SCR_PlayerSpawnPoint`, which is for making players spawn points. It has the handling for players in vehicles as well.

File: Scripts/Game/ODIN/Editor/Containers/Actions/ContextActions/ODIN_TeleportToSquadmateContextAction.c [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+
+		//if (selectedPlayer.IsInVehicle())
+		//{
+			// try to put hoveredentity inside vehicle? otherwise keep going and put it next to vic 
+			//  For now I didn't find a way to obtain the vehicle, check empty spots and directly move a player into one... Potential for future!
+		//}
+		
+		// Get world coordinates of player 
+		vector target_pos;
+		SCR_WorldTools.FindEmptyTerrainPosition(target_pos, selectedPlayer.GetOrigin(), 20);
```
Comment (Kexanone) [inline Scripts/Game/ODIN/Editor/Containers/Actions/ContextActions/ODIN_TeleportToSquadmateContextAction.c:118]: Arbitrary constants should be static member variables.

**Review by Kexanone — CHANGES_REQUESTED** [review]

File: Scripts/Game/ODIN/Editor/Containers/Actions/ContextActions/ODIN_TeleportToSquadmateContextAction.c (hunk 2 above)
Comment (Crowdedlight) [reply Scripts/Game/ODIN/Editor/Containers/Actions/ContextActions/ODIN_TeleportToSquadmateContextAction.c:65]: Woops. Missed that one. Done

File: Scripts/Game/ODIN/Editor/Containers/Actions/ContextActions/ODIN_TeleportToSquadmateContextAction.c (hunk 3 above)
Comment (Crowdedlight) [reply Scripts/Game/ODIN/Editor/Containers/Actions/ContextActions/ODIN_TeleportToSquadmateContextAction.c:74]: Just checked. That seems to work. Was unsure if the object comparison was implemented.

File: Scripts/Game/ODIN/Editor/Containers/Actions/ContextActions/ODIN_TeleportToSquadmateContextAction.c (hunk 4 above)
Comment (Crowdedlight) [reply Scripts/Game/ODIN/Editor/Containers/Actions/ContextActions/ODIN_TeleportToSquadmateContextAction.c]: Thanks for the hint! That did the trick.

**Review by Kexanone — APPROVED** [review]

**Pull request merged**